logstash-core/src/main/java/org/logstash/settings/Coercible.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.settings;
+
+import java.util.function.Predicate;
+
+public abstract class Coercible<T> extends Setting<T> {
+    public Coercible(String name, T defaultValue, boolean strict, Predicate<T> validator) {
+        super(name, strict, validator);
+
+        if (strict) {
+            T coercedDefault = coerce(defaultValue);
+            validate(coercedDefault);
+            this.defaultValue = coercedDefault;
+        } else {
+            this.defaultValue = defaultValue;
+        }
+    }
+
+    @Override
+    public void set(T value) {
+        T coercedValue = coerce(value);
+        validate(coercedValue);
+        super.set(coercedValue);
+    }
```

Comment:
The motivation for the `void set(T value)` was to express more clearly what `set` expect, a sort of contract, which in my expectations, once we have all the Settings re-implemented in Java, would make evident some type-related problems in the Setting hierarchy.
One of the thing that I tried to do is to keep a syntactic compatibility of new Java classes with previous Ruby classes, to limit changes to existing code.

If I  understand correctly your second proposal consist in defining 2 type params, one for the "read value" (`VALUE`) so that a Boolean setting would use Boolean (java type object) and is used in every "read" Setting's method.
The `ACCEPT` is instead the data type used to `set` the value, so for example could say that Boolean setting accepts and decodes String value like
```java
class Boolean extends Setting<Boolean, String>
```

That would be a nice idea, but as of today the `set` is invoked only from Ruby code and maybe that is an overkill.

I would choose the path to use `Object` as param type of the `set` method, but this creates other problems:
- the `validate` method in Setting's root class should also accept an `Object` type:  [code ref](https://github.com/elastic/logstash/pull/15679/files#diff-1c000460fe4b7a9d036450a839f0df698de1cc03bc24ce4f2a679d479bda8d91R155)
- also the `value` field in Setting root at that point miss the type and become an `Object` : [code ref](https://github.com/elastic/logstash/pull/15679/files#diff-1c000460fe4b7a9d036450a839f0df698de1cc03bc24ce4f2a679d479bda8d91R157)

I understand the actual proposition is less then optimal, but moving to Object would force to loose the type in many parts.

WDYT? Do you foresee a problem in keeping the type also in coerced set?

Reply:
@yaauie please could you review my last 🔝  comment 🙏

Reply:
As defined here, any caller of this `Coercible<T>` (or its subclasses) will (a) need to _know_ that the setting is `Coercible`, and (b) will need to use `T Coercible<T>#coerce(Object)` to transform the `Object` value into an instance of `T` that can be accepted by `Coercible<T>#set(T)`. This significantly breaks the encapsulation.

A middle-ground could be:
 - ([A](https://gist.github.com/yaauie/62782653e8ea1620809366d8f813fc18#file-java-settings-setsafely-diff-L74-L80)) renaming `Setting<T>#set(T)` -> `Setting<T>#setSafely(T)` and introducing `Setting<T>#set(Object)` with a _default_ implementation that merely does a class-cast (therefore being exceptional if given the wrong type). In this way the caller can use one method if they _know_ that they are working with the expected type, and the other to rely on coercion _without needing to introspect the setting that they are working with_.
 - ([B](https://gist.github.com/yaauie/62782653e8ea1620809366d8f813fc18#file-java-settings-setsafely-diff-L22-L36)) Implementing both `Coercible<T>#set(Object)` and `Coercible<T>#setSafely(T)` to use `Coercible<T>coerce(Object)`, and having a private shared `Coercible<T>#set(T)` that performs the always-on validation but doesn't re-coerce the value

I whipped up a quick diff representing the above [here](https://gist.github.com/yaauie/62782653e8ea1620809366d8f813fc18#file-java-settings-setsafely-diff).

Reply:
Thank's for the suggestion I've completely missed that forcing the client of `CoercibleSetting.set` to know upfront the type of the value to set is a violation of the encapsulation.
I've integrated you suggestion, the only observation about point `B` 
> having a private shared Coercible<T>#set(T)
 We can't have it because would conflict with `set(Object)` due to type erasure, we could override methods only if they have different argument lists, but for the JVM `set(Object val)` and `set(T val)` are  the same method.